refaster-support/src/main/java/tech/picnic/errorprone/refaster/matchers/IsMethodInvocationWithTwoOrMoreArgs.java

```diff
@@ -0,0 +1,27 @@
+package tech.picnic.errorprone.refaster.matchers;
+
+import static com.google.errorprone.matchers.Matchers.anyMethod;
+import static com.sun.source.tree.Tree.Kind.METHOD_INVOCATION;
+
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.matchers.Matcher;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+
+/** A matcher for method invocations with two or more arguments. */
+public final class IsMethodInvocationWithTwoOrMoreArgs implements Matcher<ExpressionTree> {
+  private static final long serialVersionUID = 1L;
+
+  /** Instantiates a new {@link IsMethodInvocationWithTwoOrMoreArgs} instance. */
+  public IsMethodInvocationWithTwoOrMoreArgs() {}
```

Comment:
Interesting, [here](url) we discussed that we want to check for two MethodInvocations instead of one method invocation wit two arguments. This is not incorrect though. I think it is nice to combine the two things into one `Matcher` and call it something like: `IsChainedMethodCallOrHasAtLeastTwoArgs`. I tried implementing this but had some troubles with the `VisitorState` in the `matches` method. 

In the `AbstractMatcherTestChecker` we only pass the `VisitorState` from the `CompilationUnit` which is not the state that we want. So we need to probably change that setup a bit. For now I'll flush the changes that I have.

Reply:
> Interesting, [here](https://github.com/PicnicSupermarket/error-prone-support/pull/url) we discussed that we want to check for two MethodInvocations instead of one method invocation wit two arguments.

Ah, that was a misunderstanding on my part, but indeed both could apply for this situation. I can look into it a bit more to implement a (combined) matcher for chained method calls as well

Reply:
Hey @rickie, few questions:
1. Could you elaborate on the `VisitorState` issue you mentioned? Which state should we be using instead?
2. Could nested method invocations also apply for this situation? So something like:
```java
m1(m2("param"))
```

Reply:
Indeed, the matcher shouldn't require >= 2 arguments: `foo().bar()` doesn't have any arguments, yet should be matched. One option is to use a `TreeScanner` (see examples in the code base) to count the number of `MethodInvocationTree`s.

IIUC the idea is to match non-trivial expressions that can't (or shouldn't) be replaced with a method reference. In this case we know that we're rewriting the expression to a `Supplier<T>`. So perhaps we're looking for a matcher of expressions which:
1. Contain more than one method invocation; or
2. Contain at least one method invocation with _one_ or more arguments.

(Invocations of a single nullary method such as `foo()`, `x.foo()` or `x.y.foo()` should not be matched for now, since in those case a method reference may be preferable.)

**That said**, expressions of the form `42 * foo()` should also be deferred.... So perhaps we should instead introduce an `IsLikelyTrivialComputation` matcher that  _does_ match `true`, `x`, `foo()`, `x.foo()`, `x.y.foo()` etc., and then have the Refaster rule use `@NotMatches(IsLikelyTrivialComputation.class)`. Then later once `MethodReferenceUsage` works properly, we can replace this matcher with a simpler `RequiresComputation` matcher and then use `@Matches(RequiresComputation.class)`.

Reply:
> In the `AbstractMatcherTestChecker` we only pass the `VisitorState` from the `CompilationUnit` which is not the state that we want. So we need to probably change that setup a bit. For now I'll flush the changes that I have.

I have a fix for this on a branch somewhere; let me try to open a PR for this ~soon.

Reply:
I filed #599.

Reply:
Nice examples @mlrprananta ! 

I checked in IntelliJ and saw that it mentions this optimization once you press `alt+enter` on the specific method. However, it doesn't highlight the method invocation to indicate you _should_ rewrite it to a method reference, which for simple cases IntelliJ does. Because for the more difficult ones it doesn't do that, I would say it is fine for now to consider these things as a non trivial computation and we can rewrite them. Probably the last three examples will not occur that often. 

So I think the fist three examples are on the "edge" of being a "trivial" computation... I'm not sure about those. The others can definitely be rewritten in the context of this PR.

Reply:
Alright, so I had a close look at this, and pushed a proposal :upside_down_face: